cpp/src/join/mixed_join_kernels.cuh

```diff
@@ -0,0 +1,331 @@
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+  std::size_t block_counter = BlockReduce(temp_storage).Sum(thread_counter);
+
+  // Add block counter to global counter
+  if (threadIdx.x == 0) atomicAdd(output_size, block_counter);
+}
+
+/**
+ * @brief Performs a join conditioned on a predicate to find all matching rows
+ * between the left and right tables and generate the output for the desired
+ * Join operation.
+ *
+ * @tparam block_size The number of threads per block for this kernel
+ * @tparam output_cache_size The side of the shared memory buffer to cache join
+ * output results
+ * @tparam has_nulls Whether or not the inputs may contain nulls.
+ *
+ * @param[in] left_table The left table
+ * @param[in] right_table The right table
+ * @param[in] join_type The type of join to be performed
+ * @param[out] join_output_l The left result of the join operation
+ * @param[out] join_output_r The right result of the join operation
+ * writes to the global output
+ * @param device_expression_data Container of device data required to evaluate the desired
+ * expression.
+ * @param[in] swap_tables If true, the kernel was launched with one thread per right row and
+ * the kernel needs to internally loop over left rows. Otherwise, loop over right rows.
+ */
+template <cudf::size_type block_size,
+          cudf::size_type output_cache_size,
+          bool has_nulls,
+          typename OutputIt1,
+          typename OutputIt2>
+__global__ void mixed_join(table_device_view left_table,
+                           table_device_view right_table,
+                           table_device_view probe,
+                           table_device_view build,
+                           null_equality compare_nulls,
+                           join_kind join_type,
+                           cudf::detail::mixed_multimap_type::device_view hash_table_view,
+                           OutputIt1 join_output_l,
+                           OutputIt2 join_output_r,
+                           cudf::ast::detail::expression_device_view device_expression_data,
+                           cudf::size_type const* matches_per_row,
+                           cudf::size_type const* join_result_offsets,
+                           bool const swap_tables)
+{
+  // Normally the casting of a shared memory array is used to create multiple
+  // arrays of different types from the shared memory buffer, but here it is
+  // used to circumvent conflicts between arrays of different types between
+  // different template instantiations due to the extern specifier.
+  extern __shared__ char raw_intermediate_storage[];
+  cudf::ast::detail::IntermediateDataType<has_nulls>* intermediate_storage =
+    reinterpret_cast<cudf::ast::detail::IntermediateDataType<has_nulls>*>(raw_intermediate_storage);
+  auto thread_intermediate_storage =
+    &intermediate_storage[threadIdx.x * device_expression_data.num_intermediates];
```

Comment:
This feels like it can all be hidden as a detail of the `expression_evaluator` type, or at least simplified to look more like CUB's pattern, e.g.,
```
    using BlockReduce = cub::BlockReduce<int, 128> BlockReduce;
    __shared__ typename BlockReduce::TempStorage temp_storage;
    auto aggregate = BlockReduce(temp_storage).Sum(thread_data);
```

so something like...
```
  using Evaluator = expression_evaluator<has_nulls, block_size>;
  extern __shared__ Evaluator::temp_storage storage;
  auto evaluator = Evaluator(storage, left_table, right_table, device_expression_data);
```

Reply:
Doesn't this CUB pattern rely on the fact that the size of `BlockReduce::TempStorage` is known at compile time based on the template parameters of the `BlockReduce`? In principle you could still do it for dynamic shared memory like we use in the `expression_evaluator`, but only if you were OK with the `expression_evaluator` always consuming all of the dynamic shared memory, which would preclude any optimizations that involve splitting dynamic shared between the evaluator and some other kernel-specific variable.

Reply:
> Doesn't this CUB pattern rely on the fact that the size of `BlockReduce::TempStorage` is known at compile time based on the template parameters of the `BlockReduce`?

It does use this, yes, but it isn't necessary. The example I gave is still using dynamic shared memory (see the `extern`), but just using the `Evaluator` typedef to simplify getting the type to use for the shared memory declaration and then hiding the cast internally.

Reply:
Discussed offline, we want to look into this but it's not urgent for this PR and can be tackled in the future. I've made a note of this in #9557.